File: ruby/bingo_board.rb
# frozen_string_literal: true

class BingoBoard
  class IllegalStateException < StandardError; end

  def initialize(width, height)
    @cells = Array.new(width) { Array.new(height) }
    @marked = Array.new(width) { Array.new(height, false) }
  end

  def define_cell(x, y, value)
    if @cells[x][y]
      raise IllegalStateException, 'Cell already defined.'
    end

    @cells.each_with_index do |row, c|
      row.each_with_index do |cell, r|
        if value == cell
          raise IllegalStateException, "#{value} already present at #{c}, #{r}"
        end
      end
    end

    @cells[x][y] = value
  end

  def mark_cell(x, y)
    unless initialized?
      raise IllegalStateException, 'Board is not initialized.'
    end

    @marked[x][y] = true
  end

  def initialized?
    @cells.all? { |row| row.all? { |cell| !cell.nil? } }
  end

  def marked?(x, y)
    @marked[x][y]
  end
end

File: php/src/BingoBoard.php
<?php

namespace Kata;

use RuntimeException;

class BingoBoard
{
    private array $cells;
    private array $marked;

    public function __construct(int $width, int $height)
    {
        $this->cells = array_fill(0, $width, array_fill(0, $height, null));
        $this->marked = array_fill(0, $width, array_fill(0, $height, false));
    }

    public function defineCell(int $x, int $y, string $value): void
    {
        if ($this->cells[$x][$y] !== null) {
            throw new RuntimeException("cell already defined");
        }

        for ($c = 0; $c < count($this->cells); $c++) {
            for ($r = 0; $r < count($this->cells[$c]); $r++) {
                if ($value === $this->cells[$c][$r]) {
                    throw new RuntimeException("$value already present at $c,$r");
                }
            }
        }

        $this->cells[$x][$y] = $value;
    }

    public function markCell(int $x, int $y): void
    {
        if (!$this->isInitialized()) {
            throw new RuntimeException("board not initialized");
        }
        $this->marked[$x][$y] = true;
    }

    public function isMarked(int $x, int $y): bool
    {
        return $this->marked[$x][$y];
    }

    public function isInitialized(): bool
    {
        foreach ($this->cells as $row) {
            foreach ($row as $col) {
                if ($col === null) {
                    return false;
                }
            }
        }
        return true;
    }
}
